### PR TITLE
New rule S6800: Component parameter type should match the route parameter type constraint

### LINKED ISSUE #8275: New rule S6800: Component parameter type should match the route parameter type constraint
S6800: https://github.com/SonarSource/rspec/pull/3213

### PR DESCRIPTION
**Pull request by sebastien-marichal**
Fixes #8275

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:52]: This is the second time within 2 weeks, that we should use the unexposed RegisterSymbolStartAction. This kind of analyzer would be much more straightforward with this:
* Register for [SymbolStartAction](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.compilationstartanalysiscontext.registersymbolstartaction?view=roslyn-dotnet-4.6.0) for classes.
* Collect the class route attributes in a list and if not empty, register for PropertyDeclarions.
* In the PropertyDeclaration do a list lookup and raise on mismatch.
 
This is the second example where we should have used such an approach.
https://github.com/SonarSource/sonar-dotnet/pull/8251#discussion_r1371343596

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.razor
Comment (antonioaversa) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.razor:12]: I would also try with an optional struct, such as `int` or `long`.
Do they work with nullable value types? 
That is: if you declare an `int?` on the property, is that compatible with both an `int` and an `int?` on the route constraint side?

Another question that would How do they combine with non-nullable reference types?
It would be worth to add such tests.

**Review by antonioaversa — CHANGES_REQUESTED** [review]
The approach and the structure of the rule is good.
I have a few educational questions and requests to test or support additional cases, as well as suggestions for code restructuring and code conventions issues to address.

**Comment by antonioaversa** [comment]
@sebastien-marichal SonarCloud identifies two code sections not fully covered:
```cs
ConstraintMapping.ContainsKey(constraintType) && ConstraintMapping[constraintType] == type.Name;
```
```cs
            .Where(attr => attr.IsKnownType(KnownType.Microsoft_AspNetCore_Components_RouteAttribute)
                           && attr.ArgumentList.Arguments.Count > 0
                           && attr.ArgumentList.Arguments[0].Expression is LiteralExpressionSyntax)
```

Some of the suggestions (`Is` extension method on `ITypeSymbol` and testing routes with `Order`) may help fixing these coverage issues.

File: analyzers/tests/SonarAnalyzer.UnitTest/MetadataReferences/NuGetMetadataReference.cs
Comment (sebastien-marichal) [reply analyzers/tests/SonarAnalyzer.UnitTest/MetadataReferences/NuGetMetadataReference.cs:56]: I will remove the default value as it is better to be specific about which version.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:74]: The property might be declared in a different file (e.g. base class). I'm not sure but I think Roslyn wasn't happy when the location of the report was in another syntax tree than the one where `c.Node` is part of. Do you have a test for such a case? If it doesn't work, you have to swap primary and secondary locations.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/AttributeSyntaxExtensions.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/AttributeSyntaxExtensions.cs:39]: Such an overload should not exists IMO. In the past we always used KnowType checks that made a quick syntactic check of the name and a semantic check afterwards. Please consult Pavel or Costin as well but I don't think we want to do syntactic-only checks. This is also

File: analyzers/rspec/cs/S6800.json
Comment (sebastien-marichal) [reply analyzers/rspec/cs/S6800.json:16]: > If the proposal to change the quickfix in the rspec PR is accepted

What do you mean by that?
I see no proposal on the rspec PR, and the file is up to date with [rspec](https://github.com/SonarSource/rspec/pull/3213/files#diff-570a5eda8552fed2ba2b5c54e8126570eeaaf02eaa5ad9be89018f60a9a6fb60).

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (sebastien-marichal) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:54]: Bad case of copy-pasting 😅

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (sebastien-marichal) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:121]: Yes you can, everything is case-insensitive in the URL.
Like `/route/{BoolParam:bool}` is the same as `/route/{BOOLparam:bOOl}`

I am adding test cases for this since I forgot to add some.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.Partial.razor.cs
Comment (sebastien-marichal) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.Partial.razor.cs:3]: We could also make this namespace more self-explanatory.

File: analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S101.json
Comment (sebastien-marichal) [reply analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S101.json:5]: I don't think dots are often used in practice, I can update the file name.

File: analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S2333.json
Comment (sebastien-marichal) [reply analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S2333.json:5]: > UPDATE: it's because ITs run in SL context, and the other half of the class is not generated, right?

Yes, this is it, I should have added a comment about it.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:117]: What about implicit conversions:
* int -> int? (Nullable conversion)
* int -> object (boxing conversion)
* int -> IComparable<int> (boxing conversion)
* int -> long (Widening conversion)

Basically any of these:
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/conversions#102-implicit-conversions

Plus according to the [docs](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-7.0#route-constraints)

> Route constraints also work with [optional parameters](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-7.0#route-parameters). In the following example, Id is required, but Option is an optional boolean route parameter.

It seems an explicit conversion from bool? to bool is also supported.

There is a [CSharpExtensions.ClassifyConversion](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.csharpextensions.classifyconversion?view=roslyn-dotnet-4.6.0) which gives you [Conversion.IsImplicit](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.conversion.isimplicit?view=roslyn-dotnet-4.6.0#microsoft-codeanalysis-csharp-conversion-isimplicit)

File: analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S6800.json
Comment (sebastien-marichal) [reply analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S6800.json:13]: Arrays should be already covered.
Or maybe you are talking about more specific cases?

I added the other cases (generics, qualified type, pointers).

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.razor
Comment (sebastien-marichal) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.razor:23]: Those are not supported in Blazor (thankfully in a way).
But it might be the case at some point.

For info, the application (Blazor server) does not even start:
> ArgumentException: Unsupported constraint 'range(0, 100)' in route '/counter/{InitialCount:int:range(0, 100)}'.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (sebastien-marichal) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:142]: Nice, I was looking for something like this!

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.razor
Comment (sebastien-marichal) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.razor:12]: > Do they work with nullable value types?

Yes, see [Query strings](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-7.0#query-strings). Although it is about query string it is still relevant for route param as well, especially this part:
> Component parameters supplied from the query string support the following types:
>
> - bool, DateTime, decimal, double, float, Guid, int, long, string.
> - Nullable variants of the preceding types.
> - Arrays of the preceding types, whether they're nullable or not nullable.

----

> That is: if you declare an int? on the property, is that compatible with both an int and an int? on the route constraint side?

The `?` on the route constraint side only specifies that the route parameter is optional. See in this section, [Route parameters](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-7.0#route-parameters). It is in the end not a constraint.
In this case, when mapping to non-nullable value, the property will be its default (e.g: 0 for int), for a nullable the value will be `null`.

----

> Another question that would How do they combine with non-nullable reference types?

As stated in the previous answer, it will use the default of the type.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (sebastien-marichal) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs]: I added more tests for implicit, boxing, widening, and also enumeration conversion.

> It seems an explicit conversion from bool? to bool is also supported.

Optional parameters are not nullable, there are about the route parameter being optional when resolving the route. If the optional route parameter is missing, the default value of the property type is used.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (sebastien-marichal) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs]: I added one test to check this behavior when the route is defined in one file and the property in another.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (sebastien-marichal) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.cs:21]: Blazor [RouteAttribute](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.routeattribute?view=aspnetcore-7.0) only contains the template parameter.
I will add the cases with raw string literal.

File: analyzers/rspec/cs/S6800.json
Comment (antonioaversa) [reply analyzers/rspec/cs/S6800.json:16]: Oopsie, I forgot to the release the changes on the RSPEC PR.
It's done now. [Here](https://github.com/SonarSource/rspec/pull/3213#discussion_r1383581729) is the comment. Apologies.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/AttributeSyntaxExtensions.cs
Comment (sebastien-marichal) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/AttributeSyntaxExtensions.cs]: I revert the change as I don't need it anymore.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:222]:
```suggestion
    private readonly struct record RouteParameter(string Constraint, string FromRoute, Location RouteParamLocation);
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:224]:
```suggestion
    private readonly struct record SupportedType(KnownType ConstraintKnownType)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (sebastien-marichal) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:240]: I have multiple route that use the same property `EnumParam`:

```razor
@page "/route/4/{EnumParam:int}" @* Compliant *@
@page "/route/5/{EnumParam:long}" @* Compliant *@
@page "/route/6/{EnumParam:guid}" @* Noncompliant *@
```

Since it is confusing, I will make it more explicit.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (sebastien-marichal) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:207]: This is the only way I found to keep the `[]` and `*` in the type name without doing it manually.
Otherwise, with `GetName`, I would always end with the underlying type.
For instance:
- `int[]` becomes `int`
- `int*` becomes `int`

Using the `ToString` was the only way I found it. Based on my test, the results were expected, but I might have missed a test case.

If you know a better approach, I am all ears.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:41]: Possible optimization:
* Move `ConstraintMapping` to compilation start and capture it.
* Replace the value `new SupportedType(..)` by `Func<ITypeSymbol, bool>` (Implemented like `Matches`)

In the CompliationStartAction:
```cs
var constraintMapping = new Dictionary<string, Func<ITypeSymbol, bool>>(StringComparer.InvariantCultureIgnoreCase)
{
    { "bool", MatchFunction(KnownType.System_Boolean),
};

Func<ITypeSymbol, bool> MatchFunction(KnownType knownType)
    => MatchFunction(cc.Compilation, cc.Compilation.GetTypeByMetadataName(knownType));

static Func<ITypeSymbol, bool> MatchFunction(Compilation compilation, ITypeSymbol targetType)
    => (ITypeSymbol expressionType) =>
{
// see the old implementation
}
```

That way, 
* you do not pass the compilation to `match`
* you only call `GetTypeByMetadataName` once at the compilation start

But you have to pass the dictionary around. I would consider it, but leave it as is if it is too cumbersome.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:131]: `.Where(x=> x.StartsWith("{") && x.EndsWith("}"))`
Non-blocking: can be done in a following PR.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.Partial.razor.cs
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterTypeShouldMatchRouteTypeConstraint.Partial.razor.cs:4]: As well, yes, and even better.

File: analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S6800.json
Comment (antonioaversa) [reply analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S6800.json:13]: I may have missed the arrays in the previous version of the code. All good now.

File: analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S2333.json
Comment (antonioaversa) [reply analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S2333.json:5]: Yeah, it would be helpful. I have added it in [the PR of S6798](https://github.com/SonarSource/sonar-dotnet/issues/8275)https://github.com/SonarSource/sonar-dotnet/issues/8275, that is based on this PR.
```cs
// FN: Razor generation disabled in SL context, so in ITs too
```

I can do it in the context of https://github.com/SonarSource/sonar-dotnet/issues/8275, after this PR is merged, so that we don't create conflict issues on the rebase of 8275.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs:41]: I would propose to have this in a dedicated PR, and merge this PR without this optimization, so that we can also unblock https://github.com/SonarSource/sonar-dotnet/issues/8275.

**Review by antonioaversa — APPROVED** [review]
The implementation looks quite mature and robust. I left some more comments, but there is nothing actually affecting the functionality or the rule, so nothing really blocking the delivery of this PR.
So I propose to merge it and, if needed, have follow-up PRs with the improvements proposed. That also includes the optimization proposed by @martin-strecker-sonarsource on registration.

File: analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S101.json
Comment (antonioaversa) [reply analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S101.json]: Ok, I will do it on the PR of S6798 ITs as well, once this is merged, and that PR is rebased.

File: analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S2333.json
Comment (antonioaversa) [reply analyzers/its/expected/BlazorSample/BlazorSample--net7.0-S2333.json:5]: The change was not done in https://github.com/SonarSource/sonar-dotnet/issues/8275 to prioritize delivery. This point has been added to the ["S6800 leftovers" issue](https://github.com/SonarSource/sonar-dotnet/issues/8367).

**Pull request merged**
